source/logic.rst

```diff
@@ -0,0 +1,1223 @@
+remains within :math:`\varepsilon` of :math:`a`,
+that is, there is a number :math:`N` such that for every
+:math:`n \ge N`, :math:`| s_n - a | < \varepsilon`.
+In Lean, we can render this as follows:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+    -- END
+
+In this section, we'll establish some properties of convergence.
+
+But first, we will discuss three tactics for working equality
+that will prove useful.
+The first, the ``ext`` tactic,
+gives us a way of proving that two functions are equal.
+Let :math:`f(x) = x + 1` and :math:`g(x) = 1 + x`
+be functions from reals to reals.
+Then, of course, :math:`f = g`, because they return the same
+value for every :math:`x`.
+The ``ext`` tactic enables us to prove an equation between functions
+by proving that their values are the same
+at all the values of their arguments.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    example : (λ x y : ℝ, (x + y)^2) = (λ x y : ℝ, x^2 + 2*x*y + y^2) :=
+    by { ext, ring }
+    -- END
+
+The second tactic, the ``congr`` tactic,
+allows us to prove an equation between two expressions
+by reconciling the parts that are different:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    example (a b : ℝ) : abs a = abs (a - b + b) :=
+    by  { congr, ring }
+    -- END
+
+Here the ``congr`` tactic peels off the ``abs`` on each side,
+leaving us to prove ``a = a - b + b``.
+
+Finally, the ``convert`` tactic is used to apply a theorem
+to a goal when the conclusion of the theorem doesn't quite match.
+For example, suppose we want to prove ``a < a * a`` from ``1 < a``.
+A theorem in the library, ``mul_lt_mul_right``,
+will let us prove ``1 * a < a * a``.
+One possibility is to work backwards and rewrite the goal
+so that it has that form.
+Instead, the ``convert`` tactic lets us apply the theorem
+as it is,
+and leaves us with the task of proving the equations that
+are needed to make the goal match.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    example {a : ℝ} (h : 1 < a) : a < a * a :=
+    begin
+      convert (mul_lt_mul_right _).2 h,
+      { rw [one_mul] },
+      exact lt_trans zero_lt_one h
+    end
+    -- END
+
+This example illustrates another useful trick: when we apply an
+expression with an underscore
+and Lean can't fill it in for us automatically,
+it simply leaves it for us as another goal.
+
+The following shows that any constant sequence :math:`a, a, a, \ldots`
+converges.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    variable (a : ℝ)
+
+    -- BEGIN
+    theorem converges_to_const : converges_to (λ x : ℕ, a) a :=
+    begin
+      intros ε εpos,
+      use 0,
+      intros n nge, dsimp,
+      rw [sub_self, abs_zero],
```

Comment:
Maybe mention this will become easier at some point (after `simp` is explained).

Reply:
O.k. You should tell me when you think it is appropriate to introduce `simp`. Last time I tried, you told me that it was too early.